File: bindings/wasm/hierarchies_wasm/examples/src/07_revoke_accreditation_to_accredit.ts
// Copyright 2020-2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { Federation, FederationProperty, HierarchiesClient, PropertyName, PropertyValue } from "@iota/hierarchies/node";
import assert from "assert";
import { randomBytes } from "crypto";
import { getFundedClient } from "./util";

/**
 * Demonstrate how to revoke an accreditation to accredit to a Property.
 *
 * In this example we connect to a locally running private network, but it can
 * be adapted to run on any IOTA node by setting the network and faucet
 * endpoints.
 */
export async function revokeAccreditationToAccredit(client?: HierarchiesClient) {
    console.log("\nRunning revoke accreditation to accredit example");

    // Get the client instance
    const hierarchies = client ?? (await getFundedClient());

    // Create a new federation
    const { output: federation }: { output: Federation } = await hierarchies.createNewFederation().buildAndExecute(
        hierarchies,
    );
    console.log("\n✅ Federation created successfully!");
    console.log("Federation ID: ", federation.id);

    // Federation property name
    const propertyName = new PropertyName(["Example LTD"]);

    // Federation property value
    const value = PropertyValue.newText("Hello");

    // Add the Property to the federation
    await hierarchies
        .addProperty(federation.id, new FederationProperty(propertyName).withAllowedValues([value]))
        .buildAndExecute(hierarchies);
    console.log(`\n✅ Property ${propertyName.dotted()} added successfully`);

    // A receiver is an account that will receive the accreditation
    const receiver = "0x" + randomBytes(32).toString("hex");

    // Property
    const property = new FederationProperty(propertyName).withAllowedValues([PropertyValue.newText("Hello")]);

    // Let us issue an accreditation to accredit to the Property
    await hierarchies
        .createAccreditationToAccredit(federation.id, receiver, [property])
        .buildAndExecute(hierarchies);
    console.log(`\n✅ Accreditation to accredit issued successfully for ${receiver}`);

    // Check if the receiver has the permission to accredit
    let accreditationsToAccredit = await hierarchies.readOnly().getAccreditationsToAccredit(federation.id, receiver);
    assert(accreditationsToAccredit.accreditations.length > 0, "Receiver should have permission to accredit");
    console.log("\n✅ Accreditation to accredit found for receiver");

    // Revoke the accreditation
    const permissionId = accreditationsToAccredit.accreditations[0].id;

    await hierarchies
        .revokeAccreditationToAccredit(federation.id, receiver, permissionId)
        .buildAndExecute(hierarchies);
    console.log("\n✅ Accreditation to accredit revoked successfully");

    // Check if the accreditation was revoked
    accreditationsToAccredit = await hierarchies.readOnly().getAccreditationsToAccredit(federation.id, receiver);
    assert(accreditationsToAccredit.accreditations.length === 0, "Accreditation not revoked for receiver");
    console.log("\n✅ Accreditation successfully revoked for receiver");
}

File: hierarchies-rs/examples/07_revoke_accreditation_to_accredit.rs
// Copyright 2020-2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::collections::HashSet;

use anyhow::Context;
use hierarchies::core::types::Federation;
use hierarchies::core::types::property::FederationProperty;
use hierarchies::core::types::property_name::PropertyName;
use hierarchies::core::types::property_value::PropertyValue;
use hierarchies::core::types::timespan::Timespan;
use hierarchies_examples::get_funded_client;
use iota_sdk::types::base_types::ObjectID;
use product_common::core_client::CoreClient;

/// Demonstrate how to issue an accreditation to accredit to a Property.
///
/// In this example we connect to a locally running private network, but it can
/// be adapted to run on any IOTA node by setting the network and faucet
/// endpoints.
///
/// See the following instructions on running your own private network
/// https://github.com/iotaledger/iota/blob/develop/docs/content/developer/getting-started/connect.mdx

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // Get the client instance
    let hierarchies_client = get_funded_client().await?;

    // Create new federation
    let federation = hierarchies_client
        .create_new_federation()
        .build_and_execute(&hierarchies_client)
        .await?;

    // Federation ID
    let federation_id = *federation.output.id.object_id();

    // Trusted property name
    let property_name = PropertyName::from("Example LTD");

    // Trusted property value
    let value = PropertyValue::Text("Hello".to_owned());

    let allowed_values = HashSet::from_iter([value]);

    println!("Adding Property");

    // Add the Property to the federation
    hierarchies_client
        .add_property(federation_id, property_name.clone(), allowed_values.clone(), false)
        .build_and_execute(&hierarchies_client)
        .await
        .context("Failed to add Property")?;

    println!("Added Property");

    // A receiver is an account that will receive the accreditation
    let receiver = ObjectID::random();

    // Properties
    let properties = FederationProperty {
        name: property_name.clone(),
        allowed_values,
        shape: None,
        allow_any: false,
        timespan: Timespan::default(),
    };

    // Let us issue an accreditation to accredit to the Property
    hierarchies_client
        .create_accreditation_to_accredit(federation_id, receiver, vec![properties.clone()])
        .build_and_execute(&hierarchies_client)
        .await
        .context("Failed to issue accreditation to accredit")?;

    // Issue an accreditation to the original account
    hierarchies_client
        .create_accreditation_to_accredit(
            federation_id,
            hierarchies_client.sender_address().into(),
            vec![properties],
        )
        .build_and_execute(&hierarchies_client)
        .await
        .context("Failed to issue permission to attest")?;

    println!("Issued accreditation to accredit");

    println!("Checking if the receiver has the accreditation to accredit");
    // Check if the receiver has the permission to accredit
    let can_accredit = hierarchies_client.is_accreditor(federation_id, receiver).await?;
    assert!(can_accredit);

    // Revoke the accreditation
    let accreditations = hierarchies_client
        .get_accreditations_to_accredit(federation_id, receiver)
        .await
        .context("Failed to find accreditation to accredit")?;

    let accreditation_id = accreditations.accreditations[0].id.object_id();

    hierarchies_client
        .revoke_accreditation_to_accredit(federation_id, receiver, *accreditation_id)
        .build_and_execute(&hierarchies_client)
        .await
        .context("Failed to revoke accreditation to accredit")?;

    // Check if the accreditation was revoked
    let federation: Federation = hierarchies_client.get_federation_by_id(federation_id).await?;

    println!("Federation: {federation:#?}");

    // Check if the receiver has the accreditation to accredit
    let can_accredit = federation.governance.accreditations_to_accredit.get(&receiver).unwrap();

    assert!(can_accredit.accreditations.is_empty());
    Ok(())
}
